### PR TITLE
Nr linking meta

### LINKED ISSUE #45: Add NR-LINKING metadata to message when local log decoration is enabled
[NOTE]: # ( ^^ Provide a general summary of the request in the title above. ^^ )

## Summary

When an application logs at a rate which exceeds the agent's ability to keep up, an external log forwarder is the recommended path for clients. Because they still want to have their log data associated to their APM applications, *Local Log Decoration* provides built-in support for retaining that functionality.

*Local Log Decoration* modifies the log messages with a standardized blob which can be extracted by the Logging product to associate data appropriately. In the most direct form, the blob goes into the log messages which are then written to file-system, console or other sources for later forwarding to New Relic.

The format for the decoration blob is as follows:

```
NR-LINKING|{entity.guid}|{hostname}|{trace.id}|{span.id}|{entity.name}|
```

`entity.name` **MUST** be URI encoded (in particular any embedded `|` characters **MUST** be represented as `%7C` to avoid parsing problems). `entity.name` **MAY** also be UTF-8 encoded, if platform defaults are unavailable.

If outside of a transaction, the `trace.id` and `span.id` portions are excluded, but the delimiters **MUST** remain to simpify extraction in the backend.

For some agents and logging frameworks it may be impossible or undesirable to directly manipulate the messages in flight. When this is the case, adding the blob to contextual data which can be formatted into the log **MAY** be used as a fallback position. Since this requires user configuration, though, we prefer the direct approach where possible.

## Additional Context
We need to refactor the enricher to be clear on its intentions now that we have 4 uses cases in there:
 * log enrichment(no metrics, forwarding, local decorating)
 * log forwarding(linking metadata in json + adding to log aggregator)
 * local decoration(appending NR-LINKING piped metadata to message key)
 * log metrics(adding counts around log lines and log lines by level)

### PR DESCRIPTION
**Pull request by bizob2828**
## Proposed Release Notes
 * Added NR-LINKING metadata when local decorating is enabled
 * Cleaned up enricher to appropriately handle all use cases(log enriching, log forwarding, log decorating, supportability metrics on load)
 * Removed passing in priority to log aggregator as that has to get calculated later(See https://github.com/newrelic/node-newrelic/issues/1172)

## Links
Closes #45

### CODE REVIEW

File: packages/winston-log-enricher/lib/createFormatter.js [hunk 1]
```diff
@@ -13,86 +13,192 @@ module.exports = function createFormatter(newrelic, winston) {
...
-      // sampled.
-      const priority = segment && segment.transaction && segment.transaction.priority
-      newrelic.agent.logs.add(info, priority)
+    if (isLocalDecoratingEnabled(config)) {
+      const metadata = newrelic.getLinkingMetadata(true)
+      info.message += formatLinkingMetadata(metadata)
+    } else if (isLogFowardingEnabled(config, newrelic)) {
+      const metadata = newrelic.getLinkingMetadata(true)
+      reformatLogLine(info, metadata, newrelic)
+      newrelic.agent.logs.add(info)
```
Comment (garbados) [inline packages/winston-log-enricher/lib/createFormatter.js:32]: Doesn't this require a priority? Or did we determine it doesn't after all?

File: packages/winston-log-enricher/lib/createFormatter.js [hunk 2]
```diff
@@ -13,86 +13,192 @@ module.exports = function createFormatter(newrelic, winston) {
...
-      // priority is usually set at the end of transactions, DT being
-      // a notable exception. Otherwise, matching log priority to the
-      // transaction priority can increase the odds of both being
-      // sampled.
-      const priority = segment && segment.transaction && segment.transaction.priority
-      newrelic.agent.logs.add(info, priority)
+    if (isLocalDecoratingEnabled(config)) {
+      const metadata = newrelic.getLinkingMetadata(true)
+      info.message += formatLinkingMetadata(metadata)
+    } else if (isLogFowardingEnabled(config, newrelic)) {
```
Comment (garbados) [inline packages/winston-log-enricher/lib/createFormatter.js:29]: Does this mean that if log decoration is enabled and log forwarding is enabled, then logs will be decorated but never forwarded? Is that a valid state?

Comment (bizob2828) [reply packages/winston-log-enricher/lib/createFormatter.js:29]: they are mutually exclusive features.  Take a look at the ruby docs which will be exactly like ours https://docs.newrelic.com/docs/logs/logs-context/configure-logs-context-ruby/.  there are 3 use cases

Comment (garbados) [reply packages/winston-log-enricher/lib/createFormatter.js:29]: I see. Thanks.

File: packages/winston-log-enricher/lib/createFormatter.js (hunk 1 above)
Comment (jordigh) [reply packages/winston-log-enricher/lib/createFormatter.js:32]: `newrelic.agent.logs` is now a custom class that handles the priority on its own.

https://github.com/newrelic/node-newrelic/pull/1174/files#diff-44e3efd3723af2a04792734d1602a1816d5e2d03708721c76e586eb8182b25ccR16

File: packages/winston-log-enricher/lib/createFormatter.js [hunk 3]
```diff
@@ -13,86 +13,192 @@ module.exports = function createFormatter(newrelic, winston) {
...
+ * Adds supportability metric based on agent config
+ *  1. if log enricher `Supportability/ExternalModules/WinstonLogEnricher`
+ *  2. if application logging `Supportability/Logging/Nodejs/winston/enabled`
+ *
+ * @param {Agent} agent new relic agent instance
+ */
+function createModuleUsageMetric(agent, config) {
+  if (isLogEnricher(config)) {
+    agent.metrics
+      .getOrCreateMetric('Supportability/ExternalModules/WinstonLogEnricher')
```
Comment (jordigh) [inline packages/winston-log-enricher/lib/createFormatter.js:52]: Remind me, why did we pick this prefix for this metric?

**Review by garbados — DISMISSED** [review]
I love this. So clean and organized. LGTM!

**Comment by bizob2828** [comment]
I'm going to add a few comments suggested by @michaelgoin then I'll need a reapproval

**Review by jordigh — APPROVED** [review]

**Review by garbados — APPROVED** [review]

**Pull request merged**